In [1]:
import collections
from sklearn import preprocessing
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import csv as csv
import shutil
import os
import json as json
import tensorflow as tf

/anaconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: compiletime version 3.5 of module 'tensorflow.python.framework.fast_tensor_util' does not match runtime version 3.6
  return f(*args, **kwds)
/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters


<h3>Converting json data to csv</h3>

In [2]:
outfile = open("review_stars.tsv", 'w')
sfile = csv.writer(outfile, delimiter ="\t", quoting=csv.QUOTE_MINIMAL) 
sfile.writerow(['business_id','stars', 'text'])
with open('./yelp_dataset/yelp_academic_dataset_review.json', encoding='utf-8') as f: 
    for line in f:
        row = json.loads(line)
        # some special char must be encoded in 'utf-8' 
        sfile.writerow([row['business_id'], row['stars'], (row['text']).encode('utf-8')])
outfile.close()

In [3]:
outfile2 = open("business_dataset.tsv", 'w')
sfile2 = csv.writer(outfile2, delimiter ="\t", quoting=csv.QUOTE_MINIMAL) 
sfile2.writerow(['business_id','review_count', 'stars'])
with open('./yelp_dataset/yelp_academic_dataset_business.json', encoding='utf-8') as f2: 
    for line in f2:
        row = json.loads(line)
        # some special char must be encoded in 'utf-8' 
        sfile2.writerow([row['business_id'], row['review_count'],  row['stars']])
outfile2.close()

<h3>Reading file into data frame</h3>

In [4]:
df1 = pd.read_csv('review_stars.tsv', delimiter ="\t", encoding="utf-8",na_values=['NA','?'])
df2 = pd.read_csv('business_dataset.tsv', delimiter ="\t", encoding="utf-8",na_values=['NA','?'])

In [5]:
dfa = df1[0:50000]
dfb = df2[0:50000]
dfa

business_id  stars  \
0      iCQpiavjjPzJ5_3gPD5Ebg      2   
1      pomGBqfbxcqPv14c3XH-ZQ      5   
2      jtQARsP6P-LbkyjbO1qNGg      1   
3      elqbBhBfElMNSrjFqW3now      2   
4      Ums3gaP2qM3W1XcA5r6SsQ      5   
5      vgfcTvK81oD4r50NMjU2Ag      1   
6      AxeQEz3-s9_1TyIo-G7UQw      5   
7      zdE82PiD6wquvjYLyhOJNA      5   
8      EAwh1OmG6t6p3nRaZOW_AA      4   
9      atVh8viqTj-sqDJ35tAYVg      2   
10     yFumR3CWzpfvTH2FCthvVw      5   
11     UBv8heCQR0RPnUQG0zkXIQ      1   
12     hdgYnadxg0GANhWOJabr2g      5   
13     yEOu75XjwczngvWWlr0M_A      4   
14     f5O7v_X_jCg2itqacRfxhg      5   
15     XqNDr54eLDLRfZwo4l4dVA      4   
16     0bjFYstj8USMzEV4ZQldjA      5   
17     q0n4I-zqiI47xispOqc1lA      5   
18     alG1fb1kl2vmT8s34jbbHg      5   
19     gZGsReG0VeX4uKViHTB9EQ      5   
20     -bd4BQcl1ekgo7avaFngIw      5   
21     u73j2VQ3TGWdMO-AG7MABw      2   
22     VyVIneSU7XAWgMBllI6LnQ      5   
23     3Mx4renubPRnjHUw1n2UkA      1   
24     d4z4gjdhQYs-WOVClISf_A      5   
25     f-v1fvtnbdw_QQRsCnwH-g      1   
26     yz66FIUPDKGhILDWzRLeKg      1   
27     6qDRqHWqf0EeHupSUEBfKg      5   
28     7AlULGZI1pHt0imODsqdkg      5   
29     th4pZ5LkoIIkb1Vmu_m2DA      4   
...                       ...    ...   
49970  fJ7uw4dDFIga57868Oz1EA      5   
49971  UidScb9HFDSD3NBw8MLuGw      4   
49972  C63sF-wG4f1doMIMv2qvOg      4   
49973  PEq9E-VCvYw1JdZf0MdvAA      5   
49974  Z4dpa57ch2R8ya1baF36iA      1   
49975  v1C8WQR6iB0WW0W48T5pqA      2   
49976  r_BrIgzYcwo1NAuG9dLbpg      4   
49977  kOFDVcnj-8fd3doIpCQ06A      4   
49978  UxWH8zRYIBgs6Q2oykvRdw      2   
49979  OedNdVrDavAlUBV7d9oKXw      4   
49980  oJSa5HCiZXKLXxggQmecEQ      4   
49981  O-b5osM0NO4f31dp6_DatQ      3   
49982  hcHaWzM5wfdAbCWs4zmmqg      4   
49983  2Q89PhkcyT9uZnQORmqMKg      3   
49984  G_05ZI4Vpl4bxjjZR11jEQ      4   
49985  EQCd_CUFrK_gSARL7s01fw      4   
49986  5T6kFKFycym_GkhgOiysIw      5   
49987  ZtQ47KIMcwy3l8LToVKeVQ      4   
49988  lHazo7qKEcSc62PMUW3oaw      4   
49989  l1CGkZpxkFrdrIz1Xl-bsA      5   
49990  tuh6HTHEXvBAN65iioxOog      1   
49991  hVcnr0TMi500pJvghwnSOQ      4   
49992  OgJ0KxwJcJ9R5bUK0ixCbg      2   
49993  U2W22Vy1oSs4G7W-2vdQ4w      1   
49994  N93EYZy9R0sdlEvubu94ig      4   
49995  q0tMKr-hzS4UT0RArdZfzQ      5   
49996  GqSA5t_OD1uxDZMwVLpoXA      3   
49997  CogdxyYsTtyHyTMuj0RgFQ      3   
49998  p7UMT_NcYbMKQ-TPZ0vp3Q      5   
49999  JVCkziWkIPbd-6uY4kKZSQ      5   

                                                    text  
0      b"The pizza was okay. Not the best I've had. I...  
1      b'I love this place! My fiance And I go here a...  
2      b'Terrible. Dry corn bread. Rib tips were all ...  
3      b'Back in 2005-2007 this place was my FAVORITE...  
4      b'Delicious healthy food. The steak is amazing...  
5      b'This place sucks. The customer service is ho...  
6      b"If you like Thai food, you have to try the o...  
7      b'AMAZING!!!\n\n I was referred here by a frie...  
8      b'Ribs = amazing\n2 hour wait time= not so ama...  
9      b"Food is pretty good, not gonna lie. BUT you ...  
10     b'I have been an Emerald Club member for a num...  
11     b'The score should be negative. Its HORRIBLE. ...  
12     b'I went there twice and I am pretty happy wit...  
13     b'Great for the price, only $25 for long hair!...  
14     b"Sansotei serves some top notch ramen. They t...  
15     b"Of the several places to get bagels in Oakla...  
16     b'Cool little place, nice atmosphere and staff...  
17     b'Delicious, friendly staff, cool atmosphere, ...  
18     b'Great haircut, cool atmosphere, and really n...  
19     b'Finally! After trying many Mexican restauran...  
20     b"One of the best breakfast places I've been t...  
21     b"Food was alright, the service was HORRIBLE! ...  
22     b'DELICIOUS! forgot to take a before picture b...  
23     b'If I could leave zero stars I would. This ru...  
24     b'Removed several pine trees from our house in...  
25 

In [6]:
dfb

business_id  review_count  stars
0      Apn5Q_b6Nz61Tq4XzPdf9A            24    4.0
1      AjEbIBw6ZFfln7ePHha9PA             3    4.5
2      O8S5hYJ1SMc8fA4QBtVujA             5    4.0
3      bFzdJJ3wp3PZssNEsyU23g             8    1.5
4      8USyCYqpScwiNEb58Bt6CA             4    2.0
5      45bWSZtniwPRiqlivpS8Og            63    4.0
6      9A2quhZLyWk0akUetBd8hQ             7    4.0
7      6OuOZAok8ikONMS_T3EzXg             7    2.0
8      8-NRKkPY1UiFXW20WXKiXg            40    2.5
9      UTm5QZThPQlT35mkAcGOjg             3    3.5
10     KapTdGyGs7RK0c68Z6hhhg            12    1.5
11     tZnSodhPwNr4bzrwJ1CSbw           146    4.0
12     _c3ixq9jYKxhLUB0czi0ug           277    4.0
13     bBUMib8l6Me1ZB1_Qkezkg            19    3.5
14     o_zDURWzgt5yM5HW_xGJsg             3    4.5
15     IbTDLI02Sq50C7B-EWK1sg             4    2.5
16     r89Re4FNgVWHgBfjCVZyVw            12    3.5
17     hTzcHtk4-0QJnFUbkKpd5Q             4    4.0
18     UwIpS9UKsPiKAv1fiEYhqg             4    1.0
19     vJIuDBdu01vCA8y1fwR1OQ             3    1.5
20     YZCHr68c5aEVHz0bkq9K2g            20    5.0
21     gJ5xSt6147gkcZ9Es0WxlA             5    3.0
22     3ByGQOVgds2YEu6kzl-XEQ             3    2.5
23     5J3b7j3Fzo9ISjChmoUoUA            23    4.0
24     CeuTRtwsq6w5rztGOyNMPg            14    2.5
25     gsoSAb9t7Ar7Ww6O_3GuOg             6    3.5
26     guRyrKMBDkcHExGzwxSfvg             4    1.0
27     PMDlKLd0Mxj0ngCpuUmE5Q            25    3.5
28     65DAbjsiws2_T600pNpWFQ             3    1.0
29     6YC6CsXRrmPv_iwfvc9onA            11    3.0
...                       ...           ...    ...
49970  Oe_oK1Nx0tTR3fHQ86F70w             6    5.0
49971  5qK-Tjo3jJLHOSgvtMcmSg             4    3.0
49972  NpWI6sfUds0oajjf6_pN6Q             3    5.0
49973  fh7zE3666mWZO4eqr0X0zg             3    1.5
49974  ShonjR_sx6yBoo0vHvrFnw            10    2.5
49975  -oKRQMWUIlb2mO0ZPCOW4Q            19    3.0
49976  3Alq2Er28QVsdxGJlQgVkg            24    3.0
49977  8Ut7jLwv3UNSIkBDVOTA0Q             4    3.0
49978  2NpoauypUWF9Me8Xaf9QWA             4    2.5
49979  OuJkz4Ns-ebuNCO7A26t9w             4    4.0
49980  DSHABCXFF8oXlNNKHYvT7w             5    3.5
49981  Q87T5PA99x1Z9F00MpNdSg             3    3.0
49982  WWtUCY03C7Xpoj0z9wKMPA            15    4.0
49983  _MjIIjJJmdi6iZhAYxtbVw             6    5.0
49984  x8MADR52s3fkToFHrsAHaQ            23    3.5
49985  ZOIDOgnZmkvyYtaPtDQBtw             8    4.0
49986  dGflKG9y-cXy57ZOJalhtA             8    4.5
49987  wkWRSubb5Um4xj1Qnim-Ow            21    4.5
49988  HGhXbgeMEcg1cIZwxyNDeg           754    3.0
49989  MXVxIS00yyefEJtT4TyTIA           122    4.0
49990  oWfjdIZG3rH4t8FlF_j0sg             3    2.5
49991  xBgQ4T6mhS5VsmK9u6mBcg           239    4.0
49992  Hn0dzCxesXWmZux79y80qQ            31    4.0
49993  tyqvpbFIC764dnQh0Qpv2g            50    4.0
49994  5zHi8eQhO7sO3N15zHgkUg           208    4.0
49995  igLKtZ9aDT8911U1xZcuSg            39    4.5
49996  mb3jJcIVW83OkoYqg_XR_A             3    3.5
49997  XhgFynFD7ZT77wofWpeOag           315    3.5
49998  E3yER5W6Wsyc5ncGKb9Nzw           126    3.0
49999  GJtJ4PPDqrebaYOQQJ96sw            11    4.0

[50000 rows x 3 columns]

In [7]:
review_agg = dfa.groupby('business_id')['text'].sum()
df_review_agg = review_agg.to_frame().reset_index()
df_review_agg[0:10]

business_id                                               text
0  --6MefnULPED_I942VcFNA  b"The incredibly rude woman behind the cashier...
1  --9e1ONYQuAa-CB_Rrw7Tw  b"It was a real up and down evening.\nApps - B...
2  --Gc998IMjLn8yr-HTzGUg  b"Delicious! \n\nCame from Chicago this evenin...
3  --Y7NhBKzLTbNliMUX_wfg  b'I have hired the Pinnacle Restoration for 3 ...
4  --cZ6Hhc9F7VkKXxHMVZSQ  b'Our family of five got the Latin Combo-- ful...
5  --cjBEbXMI2obtaRHNSFrA  b'Swindled of a $5 cover per head the doorman ...
6  --g-a85VwrdZJNf0R95GcQ  b'super fresh food..great prices. ala carte an...
7  --orEUqwTzz5QKbmyYbAWw  b"Hour-long waits are apparently typical and e...
8  --qvQS4MigHPykD2GV0-zw  b'What a fantastic moving company! It was easy...
9  -01XupAWZEXbdNbxNg5mEg  b"Amazing place to relax and grab a bite/drink...

In [8]:
df_merged = pd.merge(df_review_agg, dfb, on="business_id")
df_merged[0:10]

business_id                                               text  \
0  --6MefnULPED_I942VcFNA  b"The incredibly rude woman behind the cashier...   
1  --Y7NhBKzLTbNliMUX_wfg  b'I have hired the Pinnacle Restoration for 3 ...   
2  --g-a85VwrdZJNf0R95GcQ  b'super fresh food..great prices. ala carte an...   
3  --orEUqwTzz5QKbmyYbAWw  b"Hour-long waits are apparently typical and e...   
4  -050d_XIor1NpCuWkbIVaQ  b"One of the best breakfast places in Phoenix....   
5  -0DwB6Swi349EKfbBAOF7A  b"I was very happy to find this place!  Before...   
6  -0mm8pqBSIOYZQHeo8XnkA  b'I was in the area and was looking for a quic...   
7  -1TUzO4XCmOeqi9eAGeDWA  b"Love a mechanic that gets a minor job done i...   
8  -1VaIJza42Hjev6ukacCNg  b"The food and service aren't terrible, but li...   
9  -1oGM0x0i7I88C9g9w8XUw  b'They say "A picture paints a thousand words!...   

   review_count  stars  
0            39    3.0  
1            12    5.0  
2            24    4.5  
3             3    2.5  
4           698    4.0  
5            11    3.5  
6            40    4.0  
7            14    4.5  
8           200    4.0  
9             3    5.0

In [9]:
df_merged.shape

(7974, 4)

<h3>Label Encoding</h3>

In [10]:
from sklearn import preprocessing
le = preprocessing.LabelEncoder()
encoded_star = le.fit_transform(df_merged['stars'])
encoded_star[0:50]

array([4, 8, 7, 3, 6, 5, 6, 7, 6, 8, 7, 2, 6, 8, 7, 4, 6, 4, 6, 7, 5, 7,
       4, 8, 6, 5, 5, 6, 5, 6, 5, 6, 6, 6, 5, 4, 3, 6, 3, 6, 7, 5, 5, 6,
       6, 3, 7, 0, 6, 5])

In [11]:
import sklearn.feature_extraction.text as sk_text
# vectorizer = sk_text.CountVectorizer(min_df=1, stop_words = 'english')
vectorizer = sk_text.TfidfVectorizer(
                             stop_words='english',
                             max_features = 1000,
                             min_df=2, max_df=0.5)
matrix = vectorizer.fit_transform(df_merged.text)
matrix.shape

(7974, 1000)

In [12]:
tfidf_data = matrix.toarray()
print(tfidf_data)

[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]


<b> Normalizing data using zscore </b>

In [13]:
from scipy.stats import zscore

df_merged['normalize_rc'] = zscore(df_merged['review_count'])
df_merged[0:5]

business_id                                               text  \
0  --6MefnULPED_I942VcFNA  b"The incredibly rude woman behind the cashier...   
1  --Y7NhBKzLTbNliMUX_wfg  b'I have hired the Pinnacle Restoration for 3 ...   
2  --g-a85VwrdZJNf0R95GcQ  b'super fresh food..great prices. ala carte an...   
3  --orEUqwTzz5QKbmyYbAWw  b"Hour-long waits are apparently typical and e...   
4  -050d_XIor1NpCuWkbIVaQ  b"One of the best breakfast places in Phoenix....   

   review_count  stars  normalize_rc  
0            39    3.0     -0.345096  
1            12    5.0     -0.489705  
2            24    4.5     -0.425435  
3             3    2.5     -0.537909  
4           698    4.0      3.184447

In [14]:
y = df_merged['normalize_rc'].values.reshape( len(df_merged['normalize_rc']),1)
tfidf_data = np.concatenate([tfidf_data,y],axis=1)

In [31]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(tfidf_data,encoded_star, test_size=0.20, random_state=42)
print(tfidf_data.shape)
print(encoded_star.shape)

(7974, 1001)
(7974,)


<h2>Regression</h2>

In [49]:
from tensorflow.keras.callbacks import ModelCheckpoint, TensorBoard, EarlyStopping
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Activation, Dense
from keras.utils.np_utils import to_categorical
import matplotlib.pyplot as plt
model = Sequential()

model.add(Dense(25, input_dim=tfidf_data.shape[1], activation='sigmoid'))
model.add(Dense(10, activation='sigmoid'))
model.add(Dense(1))

model.compile(loss='mean_squared_error', optimizer='adam')  

monitor = EarlyStopping(monitor='val_loss', min_delta=1e-4, patience=3, verbose=2, mode='auto') 

checkpointer = ModelCheckpoint(filepath="best_weights.hdf5", verbose=0, save_best_only=True) # save best model
model.fit(x_train, y_train, validation_data=(x_test,y_test), callbacks=[monitor,checkpointer], verbose=2, epochs=1000)  
model.load_weights('best_weights.hdf5')

Train on 6379 samples, validate on 1595 samples
Epoch 1/1000
 - 1s - loss: 26.2716 - val_loss: 15.2195
Epoch 2/1000
 - 0s - loss: 11.2302 - val_loss: 8.6531
Epoch 3/1000
 - 0s - loss: 6.9101 - val_loss: 5.7378
Epoch 4/1000
 - 0s - loss: 4.7703 - val_loss: 4.2097
Epoch 5/1000
 - 0s - loss: 3.6745 - val_loss: 3.4546
Epoch 6/1000
 - 0s - loss: 3.1610 - val_loss: 3.1235
Epoch 7/1000
 - 0s - loss: 2.9484 - val_loss: 2.9937
Epoch 8/1000
 - 0s - loss: 2.8729 - val_loss: 2.9514
Epoch 9/1000
 - 0s - loss: 2.8496 - val_loss: 2.9393
Epoch 10/1000
 - 0s - loss: 2.8430 - val_loss: 2.9351
Epoch 11/1000
 - 0s - loss: 2.8404 - val_loss: 2.9329
Epoch 12/1000
 - 0s - loss: 2.8384 - val_loss: 2.9308
Epoch 13/1000
 - 0s - loss: 2.8356 - val_loss: 2.9275
Epoch 14/1000
 - 0s - loss: 2.8311 - val_loss: 2.9225
Epoch 15/1000
 - 0s - loss: 2.8230 - val_loss: 2.9124
Epoch 16/1000
 - 0s - loss: 2.8039 - val_loss: 2.8840
Epoch 17/1000
 - 0s - loss: 2.7276 - val_loss: 2.7526
Epoch 18/1000
 - 0s - loss: 2.4913 - val

In [50]:
pred = model.predict(x_test)
from sklearn import metrics
score = np.sqrt(metrics.mean_squared_error(pred,y_test))
print("Final score (RMSE): {}".format(score))

Final score (RMSE): 1.421084487437568


# Results

<b>Combinations tried and their results: </b>

Activation: relu, sigmoid, tanh

Optimizer: adam, sgd, rmsprop

<table>
    <thead>
        <th>Activation</th>
        <th>optimizer</th>
        <th>RMSE Score</th>     
    </thead>
    <tr>
        <td>Relu</td>
        <td>adam</td>
        <td>1.482</td>     
    </tr>
    <tr>
        <td>Relu</td>
        <td>sgd</td>
        <td>1.429</td>     
    </tr><tr>
        <td>Relu</td>
        <td>rmsprop</td>
        <td>1.457</td>     
    </tr><tr>
        <td>sigmoid</td>
        <td>adam</td>
        <td>1.418</td>     
    </tr><tr>
        <td>sigmoid</td>
        <td>sgd</td>
        <td>1.707</td>     
    </tr><tr>
        <td>sigmoid</td>
        <td>rmsprop</td>
        <td>1.424</td>     
    </tr><tr>
        <td>tanh</td>
        <td>adam</td>
        <td>1.478</td>     
    </tr><tr>
        <td>tanh</td>
        <td>sgd</td>
        <td>1.462</td>     
    </tr><tr>
        <td>tanh</td>
        <td>rmsprop</td>
        <td>1.448</td>     
    </tr>
</table>
<b> Regression Models for comparision:</b>
<table>
    <thead>
        <th>Model</th>
        <th>RMSE Score</th>
    </thead>
    <tr>
        <td>Linear Regression</td>
        <td>1.506</td>
    </tr>
</table>
<img src="Regression.png">

In [51]:
# Linear Regression
from sklearn import datasets, linear_model
linreg = linear_model.LinearRegression()

# fit the model with data
linreg.fit(x_train, y_train)

# predict the response for new observations
L_pred = linreg.predict(x_test)

score = np.sqrt(metrics.mean_squared_error(L_pred,y_test))
print("Final score (RMSE): {}".format(score))

Final score (RMSE): 1.5067510024230688


<h2>Classification</h2>

In [57]:
import pandas as pd
import io
import requests
import numpy as np
from sklearn import metrics
from keras.models import Sequential
from keras.layers.core import Dense, Activation
from keras.callbacks import EarlyStopping

In [58]:
encoded_star_2D = np.reshape(encoded_star, (len(encoded_star),1))
encoded_star_2D.shape

(7974, 1)

<h3>OneHotEncoding</h3>

In [99]:
import collections
from sklearn import preprocessing
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import shutil
import os


# Encode text values to dummy variables(i.e. [1,0,0],[0,1,0],[0,0,1] for red,green,blue)
def encode_text_dummy(df, name):
    dummies = pd.get_dummies(df[name])
    for x in dummies.columns:
        dummy_name = "{}-{}".format(name, x)
        df[dummy_name] = dummies[x]
    df.drop(name, axis=1, inplace=True)

In [100]:
df_encoded_star = pd.DataFrame(encoded_star, columns=['one_hot_encoded'])
encode_text_dummy(df_encoded_star, 'one_hot_encoded')
df_encoded_star[0:5]

one_hot_encoded-0  one_hot_encoded-1  one_hot_encoded-2  one_hot_encoded-3  \
0                  0                  0                  0                  0   
1                  0                  0                  0                  0   
2                  0                  0                  0                  0   
3                  0                  0                  0                  1   
4                  0                  0                  0                  0   

   one_hot_encoded-4  one_hot_encoded-5  one_hot_encoded-6  one_hot_encoded-7  \
0                  1                  0                  0                  0   
1                  0                  0                  0                  0   
2                  0                  0                  0                  1   
3                  0                  0                  0                  0   
4                  0                  0                  1                  0   

   one_hot_encoded-8  
0                  0  
1                  1  
2                  0  
3                  0  
4                  0

In [101]:
one_hot_array=df_encoded_star.values
one_hot_array[:10]

array([[0, 0, 0, 0, 1, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 1],
       [0, 0, 0, 0, 0, 0, 0, 1, 0],
       [0, 0, 0, 1, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 1, 0, 0],
       [0, 0, 0, 0, 0, 1, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 1, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 1, 0],
       [0, 0, 0, 0, 0, 0, 1, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 1]], dtype=uint8)

In [131]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(tfidf_data,one_hot_array, test_size=0.20, random_state=42)

In [132]:
model = Sequential()
model.add(Dense(25, input_dim= tfidf_data.shape[1], activation='sigmoid')) # Hidden 1
model.add(Dense(10, activation='sigmoid')) # Hidden 2
model.add(Dense(one_hot_array.shape[1], activation='softmax')) # Output

model.compile(loss='categorical_crossentropy', optimizer='rmsprop')
monitor = EarlyStopping(monitor='val_loss', min_delta=1e-4, patience=3, verbose=2, mode='auto') 

checkpointer = ModelCheckpoint(filepath="best_weights.hdf5", verbose=0, save_best_only=True) # save best model

model.fit(x_train, y_train, validation_data=(x_test,y_test), callbacks=[monitor,checkpointer], verbose=2,epochs=1000)

Train on 6379 samples, validate on 1595 samples
Epoch 1/1000
 - 1s - loss: 2.0818 - val_loss: 1.9222
Epoch 2/1000
 - 0s - loss: 1.8945 - val_loss: 1.8901
Epoch 3/1000
 - 0s - loss: 1.8730 - val_loss: 1.8812
Epoch 4/1000
 - 0s - loss: 1.8559 - val_loss: 1.8596
Epoch 5/1000
 - 0s - loss: 1.8316 - val_loss: 1.8327
Epoch 6/1000
 - 0s - loss: 1.8004 - val_loss: 1.8054
Epoch 7/1000
 - 0s - loss: 1.7671 - val_loss: 1.7758
Epoch 8/1000
 - 0s - loss: 1.7331 - val_loss: 1.7461
Epoch 9/1000
 - 0s - loss: 1.6997 - val_loss: 1.7158
Epoch 10/1000
 - 0s - loss: 1.6675 - val_loss: 1.6975
Epoch 11/1000
 - 0s - loss: 1.6398 - val_loss: 1.6758
Epoch 12/1000
 - 0s - loss: 1.6157 - val_loss: 1.6587
Epoch 13/1000
 - 0s - loss: 1.5947 - val_loss: 1.6488
Epoch 14/1000
 - 0s - loss: 1.5773 - val_loss: 1.6401
Epoch 15/1000
 - 0s - loss: 1.5626 - val_loss: 1.6357
Epoch 16/1000
 - 0s - loss: 1.5505 - val_loss: 1.6296
Epoch 17/1000
 - 0s - loss: 1.5390 - val_loss: 1.6287
Epoch 18/1000
 - 0s - loss: 1.5290 - val_lo

In [133]:
pred_classification = model.predict(x_test)
from sklearn import metrics
score = np.sqrt(metrics.mean_squared_error(pred_classification,y_test))
print("Final score (RMSE): {}".format(score))

Final score (RMSE): 0.29073670506477356


In [134]:
# Measure accuracy
pred = model.predict(x_test)
pred = np.argmax(pred,axis=1)

y_true = np.argmax(y_test,axis=1)

score = metrics.accuracy_score(y_true, pred)
print("Final accuracy: {}".format(score))

Final accuracy: 0.34294670846394987


# Results

<b>Combinations tried and their results: </b>

Activation: relu, sigmoid, tanh

Optimizer: adam, sgd, rmsprop

<table>
    <thead>
        <th>Activation</th>
        <th>optimizer</th>
        <th>RMSE Score</th>     
    </thead>
    <tr>
        <td>Relu</td>
        <td>adam</td>
        <td>0.296</td>     
    </tr>
    <tr>
        <td>Relu</td>
        <td>sgd</td>
        <td>0.293</td>     
    </tr><tr>
        <td>Relu</td>
        <td>rmsprop</td>
        <td>0.292</td>     
    </tr><tr>
        <td>sigmoid</td>
        <td>adam</td>
        <td>0.291</td>     
    </tr><tr>
        <td>sigmoid</td>
        <td>sgd</td>
        <td>0.303</td>     
    </tr><tr>
        <td>sigmoid</td>
        <td>rmsprop</td>
        <td>0.291</td>     
    </tr><tr>
        <td>tanh</td>
        <td>adam</td>
        <td>0.294</td>     
    </tr><tr>
        <td>tanh</td>
        <td>sgd</td>
        <td>0.290</td>     
    </tr><tr>
        <td>tanh</td>
        <td>rmsprop</td>
        <td>0.292</td>     
    </tr>
</table>
BEST result with RMSE Score obtained: <b>tanh with sgd</b>, having the accuracy of <b> 0.3429 </b>
<br>
<b> Regression Models for comparision:</b>
<table>
    <thead>
        <th>Model</th>
        <th>RMSE Score</th>
        <th>Accuracy</th>
    </thead>
    <tr>
        <td>Logistic Regression</td>
        <td>1.667</td>
        <td>0.312</td>
    </tr>
    <tr>
        <td>Support Vector Machine</td>
        <td>1.827</td>
        <td>0.272</td>
    </tr>
    <tr>
        <td>K Nearest Neighbors</td>
        <td>0.412</td>
        <td>0.232</td>
    </tr>
    <tr>
        <td>Multinomial Naive Bayes</td>
        <td>1.667</td>
        <td>0.317</td>
    </tr>
</table>

<img src="Classification.png">

In [135]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(tfidf_data,encoded_star, test_size=0.20, random_state=42)
print(tfidf_data.shape)
print(encoded_star.shape)

(7974, 1001)
(7974,)


In [143]:
# Logistic Regression
from sklearn.linear_model import LogisticRegression

# instantiate the model (using the default parameters)
logreg = LogisticRegression()

# fit the model with data
logreg.fit(x_train, y_train)

# predict the response for new observations
Log_pred=logreg.predict(x_test)

score = np.sqrt(metrics.mean_squared_error(Log_pred,y_test))
print("Final score (RMSE): {}".format(score))

# Measure accuracy
score = metrics.accuracy_score(y_test, Log_pred)
print("Final accuracy: {}".format(score))

Final score (RMSE): 1.667314398688872
Final accuracy: 0.3128526645768025


In [144]:
#Support Vector Machine
from sklearn import svm
clf = svm.SVC()
clf.fit(x_train, y_train)  
svm_pred=clf.predict(x_test)

score = np.sqrt(metrics.mean_squared_error(svm_pred,y_test))
print("Final score (RMSE): {}".format(score))

# Measure accuracy
score = metrics.accuracy_score(y_test, svm_pred)
print("Final accuracy: {}".format(score))

Final score (RMSE): 1.8273436877002796
Final accuracy: 0.2727272727272727


In [145]:
#KNN
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=1)
print(knn)
knn.fit(x_train, y_train)
Knn_pred=knn.predict(x_test)  

score = np.sqrt(metrics.mean_squared_error(Knn_pred,y_test))
print("Final score (RMSE): {}".format(score))

# Measure accuracy
score = metrics.accuracy_score(y_test, Knn_pred)
print("Final accuracy: {}".format(score))

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=1, p=2,
           weights='uniform')
Final score (RMSE): 2.059004233264682
Final accuracy: 0.23260188087774294


In [146]:
# Multinomial Naive Bayes
from sklearn import preprocessing
from sklearn.naive_bayes import MultinomialNB
min_max_scaler = preprocessing.MinMaxScaler()
mnb = MultinomialNB()
mnb.fit(min_max_scaler.fit_transform(x_train), y_train)
mnb_pred = mnb.predict(min_max_scaler.fit_transform(x_test))

score = np.sqrt(metrics.mean_squared_error(mnb_pred,y_test))
print("Final score (RMSE): {}".format(score))

# Measure accuracy
score = metrics.accuracy_score(y_test, mnb_pred)
print("Final accuracy: {}".format(score))

Final score (RMSE): 1.7795424018198072
Final accuracy: 0.31786833855799373
